File: Hashing/HashingCacheApplication.java
// File Name: HashingCacheApplication.java

import java.util.Scanner;

class HashingCacheApplication {

    static class CacheEntry {
        Integer key;
        Long value;

        CacheEntry(Integer key, Long value) {
            this.key = key;
            this.value = value;
        }
    }

    private CacheEntry[] cache;
    private int size = 10;

    public HashingCacheApplication() {
        cache = new CacheEntry[size];
    }

    private int hashFunction(int key) {
        return key % size;
    }

    private long factorial(int n) {
        long result = 1;
        for(int i = 1; i <= n; i++)
            result *= i;
        return result;
    }

    public long getFactorial(int n) {
        int index = hashFunction(n);
        int start = index;

        while (cache[index] != null) {
            if (cache[index].key == n) {
                System.out.println("Cache HIT for " + n);
                return cache[index].value;
            }
            index = (index + 1) % size;
            if (index == start)
                break;
        }

        System.out.println("Cache MISS for " + n + ". Computing...");
        long result = factorial(n);
        cache[index] = new CacheEntry(n, result);
        return result;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        HashingCacheApplication app = new HashingCacheApplication();

        while (true) {
            System.out.println("\n1.Compute Factorial\n2.Exit");
            int choice = sc.nextInt();

            if (choice == 1) {
                System.out.print("Enter number: ");
                int n = sc.nextInt();
                System.out.println("Result = " + app.getFactorial(n));
            } else {
                break;
            }
        }
    }
}

File: Hashing/hashing_cache_application.py
# File Name: hashing_cache_application.py

class HashingCacheApplication:
    def __init__(self, size=10):
        self.size = size
        self.cache = [None] * size

    def hash_function(self, key):
        return key % self.size

    def factorial(self, n):
        result = 1
        for i in range(1, n + 1):
            result *= i
        return result

    def get_factorial(self, n):
        index = self.hash_function(n)
        start = index

        while self.cache[index] is not None:
            if self.cache[index]["key"] == n:
                print(f"Cache HIT for {n}")
                return self.cache[index]["value"]
            index = (index + 1) % self.size
            if index == start:
                break

        print(f"Cache MISS for {n}. Computing...")
        result = self.factorial(n)
        self.cache[index] = {"key": n, "value": result}
        return result


if __name__ == "__main__":
    app = HashingCacheApplication()

    while True:
        print("\n1.Compute Factorial\n2.Exit")
        choice = int(input("Enter choice: "))

        if choice == 1:
            n = int(input("Enter number: "))
            print("Result =", app.get_factorial(n))
        else:
            break
